### DISCUSSION TITLE
Does IDENTIFIER allow underscores?

### DISCUSSION THREAD

**Discussion by yuki-kimoto** [Ovid/Cor#69] [Q&A] [answered]
Does IDENTIFIER allow underscores <code>_</code>?

```
Corinna        ::= CLASS | ROLE
CLASS          ::= 'class' NAMESPACE DECLARATION? BLOCK
ROLE           ::= 'role' NAMESPACE ROLES? BLOCK
NAMESPACE      ::= IDENTIFIER { '::' IDENTIFIER }
DECLARATION    ::= ':abstract'? PARENT? ROLES? VERSION?
PARENT         ::= ':isa(' NAMESPACE ')'
ROLES          ::= ':does(' NAMESPACE { ',' NAMESPACE } ','? ')'
IDENTIFIER     ::= [:alpha:] {[:alnum:]}
VERSION        ::= ':version(' VERSION_NUMBER ')
VERSION_NUMBER ::= # all allowed Perl version numbers
BLOCK          ::= # Perl +/- Extras
```

https://github.com/Ovid/Cor/blob/master/rfc/grammar.md#31-class-and-role-grammar

**Comment by Ovid** [comment, marked as answer]
Yes, it allows underscores. I've fixed and pushed that. Thanks for spotting it!

**Comment by haarg** [comment]
In actual implementation, identifiers will follow the existing rules for perl identifiers.

https://perldoc.perl.org/perldata#Identifier-parsing

**Reply by yuki-kimoto** [reply]
IDENTIFIER seems to be defined as the Perl identifier defined in the above "Identifier parsing".
